Automated tests for UI demo apps using pytest and selenium

derekpierre opened the pull request.

derekpierre commented:
@KPrasch added the workaround code for the test and the app to work when using absolute paths, see https://github.com/KPrasch/nucypher/pull/5/commits/5c51aaaed3f7c2ba51ef981eeb945d005bcf1bf2. Let me know what you think.

KPrasch commented on examples/heartbeat_rest_ui/char_control_heartbeat.py:
```diff
@@ -1,10 +1,19 @@
+    # sys.path, so that we can import the current package.  This is necessary
+    # since when invoked directly, the 'current' package is not automatically
+    # imported.
+    parent_dir = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
```
Are you able to `import nucypher` here?

KPrasch replied:
If not - you can still shorten this line a bit by `from os.path import dirname, abspath`

derekpierre replied:
I can import `nucypher` but was having issues with the examples package. Thanks for the tip on shortening the line.

derekpierre commented on tests/examples/test_heartbeat_rest_ui.py:
```diff
@@ -0,0 +1,198 @@
+    def grant_callback(request):
+        # TODO Call to control client fails with some SQL issue: `no such table: keys`
+        # TODO fake for now
+        #grant_response = alice_control_test_client.put('/grant', data=request.body)
```
Having an issue with this call at the moment. I get a SQL error about `no such table: keys` when I try to use the control client.

@KPrasch ever seen this before? I'll continue to investigate but just checking.

KPrasch replied:
Hmm - I've seen something similar - but not an active Issue that I know of.

derekpierre replied:
```
File "/Users/derek/Documents/Github/repos/forks/derek/nucypher/nucypher/keystore/keystore.py", line 112, in add_policy_arrangement
    alice_key_instance = session.query(Key).filter_by(key_data=bytes(alice_pubkey_sig)).first()
  File "/Users/derek/.virtualenvs/nucypher-z-zdnWw9/lib/python3.7/site-packages/sqlalchemy/orm/query.py", line 3169, in first
    ret = list(self[0:1])
  File "/Users/derek/.virtualenvs/nucypher-z-zdnWw9/lib/python3.7/site-packages/sqlalchemy/orm/query.py", line 2961, in __getitem__
    return list(res)
  File "/Users/derek/.virtualenvs/nucypher-z-zdnWw9/lib/python3.7/site-packages/sqlalchemy/orm/query.py", line 3271, in __iter__
    return self._execute_and_instances(context)
  File "/Users/derek/.virtualenvs/nucypher-z-zdnWw9/lib/python3.7/site-packages/sqlalchemy/orm/query.py", line 3296, in _execute_and_instances
    result = conn.execute(querycontext.statement, self._params)
  File "/Users/derek/.virtualenvs/nucypher-z-zdnWw9/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 988, in execute
    return meth(self, multiparams, params)
  File "/Users/derek/.virtualenvs/nucypher-z-zdnWw9/lib/python3.7/site-packages/sqlalchemy/sql/elements.py", line 287, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "/Users/derek/.virtualenvs/nucypher-z-zdnWw9/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1107, in _execute_clauseelement
    distilled_params,
  File "/Users/derek/.virtualenvs/nucypher-z-zdnWw9/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1248, in _execute_context
    e, statement, parameters, cursor, context
  File "/Users/derek/.virtualenvs/nucypher-z-zdnWw9/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1466, in _handle_dbapi_exception
    util.raise_from_cause(sqlalchemy_exception, exc_info)
  File "/Users/derek/.virtualenvs/nucypher-z-zdnWw9/lib/python3.7/site-packages/sqlalchemy/util/compat.py", line 374, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "/Users/derek/.virtualenvs/nucypher-z-zdnWw9/lib/python3.7/site-packages/sqlalchemy/util/compat.py", line 128, in reraise
    raise value.with_traceback(tb)
  File "/Users/derek/.virtualenvs/nucypher-z-zdnWw9/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1244, in _execute_context
    cursor, statement, parameters, context
  File "/Users/derek/.virtualenvs/nucypher-z-zdnWw9/lib/python3.7/site-packages/sqlalchemy/engine/default.py", line 546, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) no such table: keys [SQL: 'SELECT keys.id AS keys_id, keys.fingerprint AS keys_fingerprint, keys.key_data AS keys_key_data, keys.is_signing AS keys_is_signing, keys.created_at AS keys_created_at \nFROM keys \nWHERE keys.key_data = ?\n LIMIT ? OFFSET ?'] [parameters: (<memory at 0x111601a08>, 1, 0)] (Background on this error at: http://sqlalche.me/e/e3q8)
_internal.py                88 ERROR    Error on request:
```

derekpierre commented:
I updated the code with `master` (updated character control parameters etc.), and still experience the same problem. The application itself works, but the test still fails in a similar way i.e. the SQL query failing. 

```
 File "/Users/derek/Documents/Github/repos/forks/derek/nucypher/nucypher/network/server.py", line 228, in consider_arrangement
    session=session,
  File "/Users/derek/Documents/Github/repos/forks/derek/nucypher/nucypher/keystore/keystore.py", line 112, in add_policy_arrangement
    alice_key_instance = session.query(Key).filter_by(key_data=bytes(alice_pubkey_sig)).first()
  File "/Users/derek/.virtualenvs/nucypher-z-zdnWw9/lib/python3.7/site-packages/sqlalchemy/orm/query.py", line 3169, in first
    ret = list(self[0:1])
  File "/Users/derek/.virtualenvs/nucypher-z-zdnWw9/lib/python3.7/site-packages/sqlalchemy/orm/query.py", line 2961, in __getitem__
    return list(res)
  File "/Users/derek/.virtualenvs/nucypher-z-zdnWw9/lib/python3.7/site-packages/sqlalchemy/orm/query.py", line 3271, in __iter__
    return self._execute_and_instances(context)
  File "/Users/derek/.virtualenvs/nucypher-z-zdnWw9/lib/python3.7/site-packages/sqlalchemy/orm/query.py", line 3296, in _execute_and_instances
    result = conn.execute(querycontext.statement, self._params)
  File "/Users/derek/.virtualenvs/nucypher-z-zdnWw9/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 988, in execute
    return meth(self, multiparams, params)
  File "/Users/derek/.virtualenvs/nucypher-z-zdnWw9/lib/python3.7/site-packages/sqlalchemy/sql/elements.py", line 287, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "/Users/derek/.virtualenvs/nucypher-z-zdnWw9/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1107, in _execute_clauseelement
    distilled_params,
  File "/Users/derek/.virtualenvs/nucypher-z-zdnWw9/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1248, in _execute_context
    e, statement, parameters, cursor, context
  File "/Users/derek/.virtualenvs/nucypher-z-zdnWw9/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1466, in _handle_dbapi_exception
    util.raise_from_cause(sqlalchemy_exception, exc_info)
  File "/Users/derek/.virtualenvs/nucypher-z-zdnWw9/lib/python3.7/site-packages/sqlalchemy/util/compat.py", line 374, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "/Users/derek/.virtualenvs/nucypher-z-zdnWw9/lib/python3.7/site-packages/sqlalchemy/util/compat.py", line 128, in reraise
    raise value.with_traceback(tb)
  File "/Users/derek/.virtualenvs/nucypher-z-zdnWw9/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1244, in _execute_context
    cursor, statement, parameters, context
  File "/Users/derek/.virtualenvs/nucypher-z-zdnWw9/lib/python3.7/site-packages/sqlalchemy/engine/default.py", line 546, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) no such table: keys [SQL: 'SELECT keys.id AS keys_id, keys.fingerprint AS keys_fingerprint, keys.key_data AS keys_key_data, keys.is_signing AS keys_is_signing, keys.created_at AS keys_created_at \nFROM keys \nWHERE keys.key_data = ?\n LIMIT ? OFFSET ?'] [parameters: (<memory at 0x1172c5888>, 1, 0)] (Background on this error at: http://sqlalche.me/e/e3q8)
```

derekpierre commented:
@KPrasch it seems like the sqlalchemy `no such table` problem above is a threaded one because we use an in memory db for the keystore during the tests for the federated ursulas fixture; see https://gehrcke.de/2015/05/in-memory-sqlite-database-and-flask-a-threading-trap/.

The federated ursulas fixture calls `make_federated_ursulas` (nucypher/utilities/sandbox/ursula.py) which uses the following constant in `nucypher/utilities/sandbox/constants.py` as the db path:
```
MOCK_URSULA_DB_FILEPATH = ':memory:'
```

I locally changed the constant to:
```
MOCK_URSULA_DB_FILEPATH = os.path.join(BASE_TEMP_DIR, 'test')
```
and the `lifecycle` test passed i.e. the `/grant` call worked.

derekpierre commented:
To be clear, I don't think this is the solution, just an illustration of the problem 😄.

derekpierre commented on nucypher/utilities/sandbox/constants.py:
```diff
@@ -114,5 +114,5 @@ def select_test_port() -> int:
 MOCK_IP_ADDRESS_2 = '10.10.10.10'
 
-MOCK_URSULA_DB_FILEPATH = ':memory:'
+MOCK_URSULA_DB_FILEPATH = os.path.join(BASE_TEMP_DIR, f'{BASE_TEMP_PREFIX}-{os.urandom(4).hex()}')  #':memory:'
```
**NOTE: I will remove this once I figure out a better solution.**

derekpierre commented:
@KPrasch Eureka! I fixed the `InvalidTag` problem. The issue was that the enrico endpoint fixture was using a default encrypting key that was different than the one used for the policy created through the UI. I needed to create a specific enrico endpoint based on the derived key in the lifecycle test.

The only outstanding item now is the MOCK_URSULA_DB_FILEPATH. I changed it to be a tmp file instead of in-memory, since there are thread-local issues as mentioned above. This seems to be the only solution I could find. Open to your thoughts here?

derekpierre commented:
I'm going to add the other demos from nucypher/nucypher-demos repos - including automated tests.

derekpierre commented:
@KPrasch I have transferred all three demo ui apps: heartbeat_demo_ui, heartbeat_rest_ui, and vehicle_data_exchange into `nucypher/examples`. They all now have automated tests associated with each of them.

When you get some time, please review (note an update from master is probably needed). The main things to keep an eye out for are:

1. We need the `chromedriver` to be on the machine running the test - I'm unsure of the best way to ensure this, any ideas?
2. There didn't seem to be a workaround for the thread-local issue I mentioned before, so `MOCK_URSULA_DB_FILEPATH` still uses a tmp file path
3. In order for the automated unit tests to interact with the `federated_ursulas` fixture, I needed to specify the port the ursulas are running on - I used an environment variable to achieve this
4. Currently demo files are created using an absolute path - which means that actually running the app, and running the unit tests cause files to be overwritten; I perform test cleanup so files get deleted - would it be better to use relative paths? Alternatively, we could just never save Alice configuration which was the only thing kept state - but keeping the state for a simple demo may be overkill. Any thoughts here? 
5. I modified the pipfile to have the local `nucypher` code be used for running the application. That way they would remain compatible with the code as the code changes, and the tests would fail if they don't get updated for incompatible changes to `nucypher`
6. I haven't run the full `nucypher` test suite with the new ui automated tests, so I'm unsure how they interact

KPrasch commented:
Phew! Still looking this over! :-)

KPrasch commented:
1. Similar to the way we handle the solidity binary, we can include a script to prepare to local environment for browser-testing.

2. Using the temporary database file path is actually more "production-ready" and useful example then the in-memory one. :+1: 

3. Perfect.

4. Hmm - Which files get overwritten? (hopefully not existing production config files :sweat_smile: ).  I think the right solution here is the one which provides the best *code example*.  I find the try/except in the beginning of the old heartbeat demo to be cumbersome, and I think I prefer to use a dedicated temp dir for each demo, and always cleanup.  Perhaps a test showing that your cleanup logic is *safe* is useful (We do this in config tests, too).

5. OK - how many new dependencies are there?  I think it's valuable to separate the testing deps from the demo deps.  Also - I think we can utilize pip install extras ie:  `pip install nucypher[demos]`.

KPrasch commented on Pipfile:
```diff
@@ -56,6 +56,15 @@ recommonmark = "*"
 # CLI
 nucypher = {editable = true,path = "."}
+# Test Demos
+dash = "==0.34.0"
```
Do I need this dependency to *run * the demo?

derekpierre replied:
I initially started off with separate pipfiles for the demos from `nucypher` itself. It seems that now both sets of pipfiles essentially have the same dependencies. Perhaps I'll just get rid of the individual demo pipfiles, and allow use of the nucypher pipfile for running the demos. I guess this is similar for David's CLI heartbeat demo.

derekpierre commented:
>Hmm - Which files get overwritten? (hopefully not existing production config files 😅 ). I think the right solution here is the one which provides the best code example. I find the try/except in the beginning of the old heartbeat demo to be cumbersome, and I think I prefer to use a dedicated temp dir for each demo, and always cleanup. Perhaps a test showing that your cleanup logic is safe is useful (We do this in config tests, too).

The demos generate folders such as:
1. `alicia-files`: stores configuration of alicia between runs, including existing known ursulas. If there is ever a problrm loading (try,catch) then it gets deleted and new ones get created. Essentially these files are never purposely deleted
2. `bob-files`: temporary files for bob  - gets deleted and recreated on subsequent runs
3. `keys`: storage of public keys for bobs - gets deleted and recreated on subsequent runs
4. `shared`: storage of policy information that is passed between alicia and bob. Also information passed from enrico to bob - gets deleted and recreated on subsequent runs
5. `data`: storage of generated data eg. DB of data that enrico produces - gets deleted and recreated on subsequent runs

So all of the above except for #1 should probably be cleaned up after the app exits. Currently, they are not and only deleted and recreated on subsequent runs. However, for the automated tests, I added the facility to clean up the files once completed. 

>OK - how many new dependencies are there? I think it's valuable to separate the testing deps from the demo deps. Also - I think we can utilize pip install extras ie: pip install nucypher[demos].

The separation is what I was initially going for to prevent the nucypher pipfile from including dependencies only required for demos, but for running tests it may just be easier to not separate them? As I mentioned there doesn't seem to be much of a difference in the dependency list between nucypher's pipfile and the demo pipfiles.
